### ISSUE TITLE
Unable to access vector of degrees of freedom (`CR`  elements on periodic domains in parallel)

### ISSUE THREAD

**Issue by bhaveshshrimali** [FEniCS/dolfinx#917]
Migrating from the corresponding discussion on [discourse (here)](https://fenicsproject.discourse.group/t/running-cr-in-parallel-error-detected-in-hdf5/2662?u=bhaveshshrimali),

Edit: I understand it is more of a `dolfin` question than a `dolfinx` query, but in any case pointers to possible workarounds would be helpful. 

TLDR:
```python3
from dolfin import *
import numpy as np

comm = MPI.comm_world 
comm_rank = MPI.rank(comm)
set_log_level(30)

parameters["ghost_mode"] = "shared_facet"

mesh = UnitSquareMesh(20,20)

def corner(x):
    return ((near(x[0],0.)) and (near(x[1],0.)))

class PeriodicBoundary(SubDomain):
    def inside(self, x, on_boundary):
        left=bool(near(x[0],0.))
        bottom=bool(near(x[1],0.))
        right=bool(near(x[0],1.))
        top=bool(near(x[1],1.))                
        return bool((left or bottom) and (not ((left and top) or (bottom and right))) and on_boundary)

    def map(self, x, y):
        if near(x[0], 1.) and near(x[1], 1.):
            y[0] = x[0] - 1.
            y[1] = x[1] - 1.
        elif near(x[0], 1.):
            y[0] = x[0] - 1.
            y[1] = x[1] 
        elif near(x[1], 1.):
            y[0] = x[0] 
            y[1] = x[1] - 1.
        else:
            y[0] = -1.
            y[1] = -1.

V = VectorFunctionSpace(mesh, "CR", 1, constrained_domain=PeriodicBoundary())
u = Function(V)
v = TrialFunction(V)
dv = TestFunction(V)

mu=1
lmbda=5

f = Expression(("4*(l + 2*m)*pi*pi*cos(2*pi*x[0])","4*(l + 2*m)*pi*pi*sin(2*pi*x[1])"), l=lmbda, m=mu, degree=5)

bc1 = DirichletBC(V, Expression(("1","0"),degree=1), corner, method="pointwise")

def eps(v):
    return sym(grad(v))
    
F = (lmbda*tr(eps(v))*tr(eps(dv))+2*mu*inner(eps(v),eps(dv)))*dx - inner(f, dv)*dx 

h=FacetArea(mesh)         
h_avg = (h('+') + h('-'))/2
F += (1.0/h_avg)*dot(jump(v),jump(dv))*dS 

solve(lhs(F)==rhs(F), u, bc1)
```
fails when run in parallel, but works fine in serial. 
Any ideas as to what could be possibly going wrong, and any possible fixes? `CR` elements work fine when run in general (`non-periodic` bcs). A suggested workaround was to debug the constrained `DofMap` construction in FEniCS [on the same thread](https://fenicsproject.discourse.group/t/running-cr-in-parallel-error-detected-in-hdf5/2662/12?u=bhaveshshrimali)

**Comment by jorgensd** [comment]
Instead of using `constrained_domain` in `FunctionSpace`, you could try to use this extension to dolfinx, [dolfinx_mpc](https://github.com/jorgensd/dolfinx_mpc). This is under active development, and has two examples of using periodic boundary conditions.

**Comment by bhaveshshrimali** [comment]
Thanks @jorgensd . I am trying to install dolfinx. For some reason, I am not able to build a singularity container from the dolfinx docker image. I will follow along the examples once I am able to install it.

**Comment by bhaveshshrimali** [comment]
Hi @jorgensd , I managed to build a singularity container with dolfinx. But when I try to build `dolfinx_mpc` it is not able to find PETSc
```
Bootstrap: docker
From: dolfinx/real:latest

%post
    apt-get -y update
    apt-get -y install software-properties-common ffmpeg curl build-essential python3 python3-pip python3-tk
    git clone https://github.com/jorgensd/dolfinx_mpc.git 
    cd dolfinx_mpc
    rm -rf build
    mkdir -p build
    cd build
    cmake -G Ninja -DCMAKE_BUILD_TYPE=Developer ../cpp/
    ninja -j3 install
    cd ../python
    python3 -m pip install . --upgrade
    ldconfig

%runscript
    exec /bin/bash -i
```
gives the error: 
```
CMake Error at /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:146 (message):
PETSc could not be found. Be sure to set PETSC_DIR as an environment
variable. (missing: PETSC_FOUND)
```
The `Dockerfile` for `dolfinx` does seem to compile PETSc. Would you happen to know what could be going wrong? 

I am not opening a new issue. If you feel this is best discussed at another forum, I would be happy to post there. 

Thanks!

**Comment by jorgensd** [comment]
> Hi @jorgensd , I managed to build a singularity container with dolfinx. But when I try to build `dolfinx_mpc` it is not able to find PETSc
> 
> ```
> Bootstrap: docker
> From: dolfinx/real:latest
> 
> %post
>     apt-get -y update
>     apt-get -y install software-properties-common ffmpeg curl build-essential python3 python3-pip python3-tk
>     git clone https://github.com/jorgensd/dolfinx_mpc.git 
>     cd dolfinx_mpc
>     rm -rf build
>     mkdir -p build
>     cd build
>     cmake -G Ninja -DCMAKE_BUILD_TYPE=Developer ../cpp/
>     ninja -j3 install
>     cd ../python
>     python3 -m pip install . --upgrade
>     ldconfig
> 
> %runscript
>     exec /bin/bash -i
> ```
> 
> gives the error:
> 
> ```
> CMake Error at /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:146 (message):
> PETSc could not be found. Be sure to set PETSC_DIR as an environment
> variable. (missing: PETSC_FOUND)
> ```
> 
> The `Dockerfile` for `dolfinx` does seem to compile PETSc. Would you happen to know what could be going wrong?
> 
> I am not opening a new issue. If you feel this is best discussed at another forum, I would be happy to post there.
> 
> Thanks!

That seems weird, as PETSc is installed with dolfin. But as the error message says, you have not set a PETSC_DIR. You need to export the path to the PETSc installation, (in docker it is /usr/local/petsc

**Comment by bhaveshshrimali** [comment]
Yes, infact I just did that and rebuilt the image. It works now. Thanks!
